introduce torch.fx.is_tracing() for gating tracing-incompatible module logic

bradleyhd opened the pull request:
Summary: Modules sometimes are written with sanity-checking logic that can be incompatible with symbolic tracing, e.g. `isinstance(t, torch.Tensor)` or things that invoke anti-control-flow behavior like `assert len(t)`. Generally it's save to ignore these during a trace, so `torch.fx.is_tracing()` can be used to gate this behavior much like with `torch.jit.is_scripting()`.

Test Plan: Manually tested on a module that had sanity-checking behavior that normally fails a trace. Please let me know where the best place to put a unit test is if desired.

Differential Revision: D23451666

jamesr66a commented:
Can the same effect be achieved via an `isinstance(x, Proxy)` check? I'm hesitant to add more global state to the system, and a mode flag would demonstrably introduce tech debt into application code (see `is_scripting` and its callsites in fbcode)

bradleyhd commented:
Off the top of my head, I think that works (attempting to verify but having some hardware issues at the moment). I hear you on the global state -- I don't love this approach either. That said, not sure how/if `isinstance(x, Proxy)` is less tech debt in app code.

A few things that crossed my mind:
- some modules' forward methods have multiple inputs that need sanity-checked, so you'd have to decide whether to gate all the sanity checking on whether or not a single input is a `Proxy`, or start writing `isinstance(x, Proxy) and isinstance(y, Proxy)`, etc
- from a user perspective it's a bit strange to see a method signature annotated like `def forward(self, xs: List[torch.Tensor])` and have code that branches on whether `xs` is a `Proxy`. This also confuses pyre on downstream lines immediately: ```Incompatible parameter type [6]: Expected `typing.Sized` for 1st positional only parameter to call `len` but got `typing.Union[List[torch.Tensor], torch.fx.proxy.Proxy]```

houseroad reviewed (commented):
Add a test?

bradleyhd commented:
added a test for kicks, but let's keep the high-level discussion going

jamesr66a commented:
@bradleyhd 
> That said, not sure how/if isinstance(x, Proxy) is less tech debt in app code.

It's not :) The ideal case is we eliminate the need for branching, but I'm not 100% sure on how we'd do that.

jamesr66a commented:
Sorry, pushed wrong button

dzhulgakov commented on torch/fx/symbolic_trace.py:
```diff
@@ -187,6 +196,9 @@ def _proxy_placeholder(name: str, delegate: DelegateBase) -> Proxy:
 def symbolic_trace(root : torch.nn.Module, delegate_class=DefaultDelegate) -> GraphModule:
+
+    global _FX_IS_TRACING
+    _FX_IS_TRACING = True
```
I'd enforce that we currently not in tracing mode already, just in case :)

bradleyhd replied:
updated. required another nesting of `try... finally`. let me know if you know of a better way 😅

dzhulgakov replied:
I think you can keep a check on whether we're tracing already here but then move setting way down next to the existing `try:` and undo it in existing `finally:`. The only place where we run the user code is `graph.output(delegate.create_arg(fn(*args)))` so we need to wrap only that

bradleyhd replied:
✅

dzhulgakov reviewed (commented):
The downside of checking the input is that `isinstance(x, Proxy)` is not sufficient. What if it's a list/dict of proxies constructed elsewhere and passed to the function? I.e. `x : List[Tensor]` can be either `Proxy` or `List[Proxy]` during tracing. In that case we need to iterate over the elements of the container in order to look for any proxies. But that might be expensive (what if my dict is a gigantic vocab with 100k elements?).

So I feel that the global flag is the least evil. I don't think we'd be able to get rid of is_tracing checks either, mostly because of `assert isinstance()` constructs.

bradleyhd committed:
introduce torch.fx.is_tracing() for gating tracing-incompatible module logic (#43971)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/43971

Modules sometimes are written with sanity-checking logic that can be incompatible with symbolic tracing, e.g. `isinstance(t, torch.Tensor)` or things that invoke anti-control-flow behavior like `assert len(t)`. Generally it's save to ignore these during a trace, so `torch.fx.is_tracing()` can be used to gate this behavior much like with `torch.jit.is_scripting()`.

Test Plan: Manually tested on a module that had sanity-checking behavior that normally fails a trace. Added two units tests.

Differential Revision: D23451666

fbshipit-source-id: 96f96e7a118332e5ebed549c28c97b5b81984974

bradleyhd commented:
closing in favor of an internal solution that doesn't affect OSS
